src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,176 @@
+            {
+                return;
+            }
+
+            var stringBuilderSymbol = WellKnownTypes.StringBuilder(startContext.Compilation);
+
+            if (stringBuilderSymbol is null)
+            {
+                return;
+            }
+
+            var sourceAppendMethod = stringBuilderSymbol
+                .GetMembers(nameof(StringBuilder.Append))
+                .OfType<IMethodSymbol>()
+                .Where(append => append.Parameters.Length == 1)
+                .FirstOrDefault(append => append.Parameters[0].Type.SpecialType == SpecialType.System_String);
+
+            if (sourceAppendMethod is null)
+            {
+                return;
+            }
+            
+            startContext.RegisterOperationAction(
+                context =>
+                {
+                    if (context.Operation is IInvocationOperation invocation)
+                    {
+                        var invokedMethod = invocation.TargetMethod;
+
+                        if (invokedMethod == sourceAppendMethod)
+                        {
+                            var argument = invocation.Arguments.FirstOrDefault();
+                            if (argument.Value is IInvocationOperation invocationExpression)
+                            {
+                                if (invocationExpression.TargetMethod == substring1ParameterMethod)
+                                {
+                                    if (IsSideEffectFree(invocationExpression.Instance)
+                                        && IsSideEffectFree(invocationExpression.Arguments[0].Value))
+                                    {
+                                        context.ReportDiagnostic(
+                                            Diagnostic.Create(
+                                                RuleReplaceOneParameter,
+                                                invocation.Syntax.GetLocation()));
+                                    }
+                                }
+                                else if (invocationExpression.TargetMethod == substring2ParameterMethod)
+                                {
+                                    context.ReportDiagnostic(
+                                        Diagnostic.Create(
+                                            RuleReplaceTwoParameter,
+                                            invocation.Syntax.GetLocation()));
+                                }
+                            }
+                        }
+                    }
+                },
+                OperationKind.Invocation);
+        }
+
+        private bool IsSideEffectFree(IOperation instance)
+        {
+            return instance.ConstantValue.HasValue
+                   || instance.Kind == OperationKind.ConstantPattern
+                   || instance.Kind == OperationKind.ArrayElementReference
+                   || instance.Kind == OperationKind.FieldReference
```

Comment:
>instance.Kind == OperationKind.FieldReference [](start = 22, length = 45)

Field reference is not always side reference free. The instance of field reference might be an invocation for example. A field reference whose instance is null or of kind `OperationKind.InstanceReference` is side effect free.

Reply:
Then I might have to exclude field references as well, but out of interest, I don't see a way to have side effects on read access to a field. Maybe you could show me an example?